text/0000-poll-drop-ready.md

```diff
@@ -0,0 +1,480 @@
+3. Each field's destructor runs, after the type's destructor has potentially
+moved them.
+
+Because of this sequence, the in projection in `poll_drop_ready` would be
+unsound. Therefore, `poll_drop_ready` cannot take self by Pin. Here is a bit of
+code demonstrating what I mean:
+
+```rust
+// If `field` implements `poll_drop_ready` and `poll_drop_ready` used `Pin`,
+// `field` would be witnessed as pinned before calling drop on `Foo`.
+struct Foo<T: Default> {
+    field: T
+}
+
+impl<T: Default> Drop for Foo<T> {
+    fn drop(&mut self) {
+        // Then, in the destructor for `Foo`, we can move `field` and re-pin it
+        // at a different location, violating the guarantees of `Pin`.
+
+        let moved_field = mem::replace(&mut self.field, T::default());
+        let pinned: Box<Pin<T>> = Box::pin(moved_field);
+    }
+}
+```
+
+However, its worth noting that a user can locally reason about whether private
+fields can be moved or accessed again, based on the APIs exposed by a type,
+and, if necessary, reconstruct a pin of those fields. [The pin documentation
+contains relevant information.][pin-docs] In other words, if your drop
+implementation is implemented correctly, you can construct a `Pin` of self
+inside your `poll_drop_ready`.
+
+## Drop order
+
+The order of calls to `poll_drop_ready` can be inferred from the rest of this
+document, but it's good to specify them explicitly.
+
+Between variables, calls to `poll_drop_ready` will occur in reverse order of
+the variable's introduction, immediately prior to the calls to that variable's
+destructor.
+
+Between fields of a type, calls to `poll_drop_ready` will occur in the textual
+order of that type's declaration, with the call for the type itself occuring
+first. This is similar to the order for calls to drop. *However*, these calls
+will occur in a loop, at the level of the type being dropped, until all calls
+to `poll_drop_ready` for that value have returned `Ready`, so they will be
+"interleaved" and concurrent in practice. The program will *not* wait for each
+field to return ready before beginning to process the subsequent field.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Like all language extensions which increase the expressiveness of the language,
+this feature will increase the complexity of Rust. In particular, this feature
+sits squarely at the intersection of async code and destructors, both areas of
+the language which involved nuanced guarantees about memory and correctness.
+Therefore, the biggest drawback of adding this feature is that it increases the
+complexity of the language.
+
+It should also be noted that it introduces silent "yield points" into async
+contexts, whereas previously all yield points were visible awaits. This is just
+an extension of the normal behavior of destructors, which already can block,
+panic, take locks, park the thread, mutate global state, and so on (as they are
+intended to do).
+
```

Comment:
IMHO one big drawback of this approach is that it doesn't solve the requirement for implementing synchronous drop as a fallback, since there are no guarantees about async drop being called.

That kind of raises the question for application developers on "when to actually implement it", or "why to implement it at all".